### PR TITLE
feat: hydra side of ui-endpoint-info interface implementation

### PR DESCRIPTION
**Pull request by bencekov**
Let me know if I should implement more unit tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,10 +45,11 @@ This charm offers integration with [kratos-operator](https://github.com/canonica
 
 The following instructions assume that you have deployed `traefik-admin` and `traefik-public` charms and related them to hydra.
 
-If you have deployed [Login UI charm](https://github.com/canonical/identity-platform-login-ui), you can configure it with hydra by providing its URL.
+If you have deployed [Login UI charm](https://github.com/canonical/identity-platform-login-ui), you can configure it with hydra by using eachother's integrations.
```
Comment (natalian98) [inline README.md:48]: Could you split and modify the `Integration with Kratos and UI` paragraph? Also add the relation names (to avoid ambiguity) and replace `relate` with `integrate`. So sth like:

### Kratos

This charm offers integration with [kratos-operator](https://github.com/canonical/kratos-operator). In order to integrate hydra with kratos, it needs to be able to access hydra's admin API endpoint.
To enable that, integrate the two charms:
```console
juju integrate kratos hydra
```

### Identity Platform Login UI

This charm offers integration with [identity-platform-login-ui-operator](https://github.com/canonical/identity-platform-login-ui-operator).
In order to integrate them, run:
```console
juju integrate hydra:ui-endpoint-info identity-platform-login-ui-operator:ui-endpoint-info
juju integrate identity-platform-login-ui-operator:hydra-endpoint-info hydra:hydra-endpoint-info
```

File: lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+import logging
+from typing import Dict, Optional
+
+from ops.charm import CharmBase, RelationCreatedEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import Application
+
+# The unique Charmhub library identifier, never change it
+LIBID = "placeholder_ID"
```
Comment (natalian98) [inline lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py:48]: Please add the `LIBID` you registered in [this PR](https://github.com/canonical/identity-platform-login-ui-operator/pull/16)

File: lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    def send_endpoint_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "endpoint": endpoint,
```
Comment (natalian98) [inline lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py:100]: Let's make sure to update the library when all endpoints are added to the databag and [this PR](https://github.com/canonical/identity-platform-login-ui-operator/pull/16) is reviewed

**Comment by natalian98** [comment]
@bencekov I suggest to convert this PR into draft until all endpoints for the `login-endpoint-info` are placed in the databag and the provider side you're working on in `identity-platform-login-ui-operator` is updated. It will be easier to review.

Opening a PR with the relation spec in the [charm-relation-interfaces](https://github.com/canonical/charm-relation-interfaces) may help to organize the work.

File: src/charm.py [hunk 4]
```diff
@@ -180,9 +191,9 @@ def _render_conf_file(self) -> str:
 
         rendered = template.render(
             db_info=self._get_database_relation_info(),
-            consent_url=join(self.config.get("login_ui_url", ""), "consent"),
-            error_url=join(self.config.get("login_ui_url", ""), "oidc_error"),
-            login_url=join(self.config.get("login_ui_url", ""), "login"),
```
Comment (natalian98) [inline src/charm.py:185]: the config option `login_ui_url` should be removed from `config.yaml`

Comment (bencekov) [reply src/charm.py:185]: done

**Comment by gruyaume** [comment]
Please write a proper  PR description

File: src/charm.py [hunk 5]
```diff
@@ -496,6 +507,18 @@ def _update_endpoint_info(self) -> None:
             scope=" ".join(SUPPORTED_SCOPES),
         )
 
+    def _get_login_ui_endpoint_info(self, key: str) -> Optional[str]:
+        login_ui_url = ""
+        if self.model.relations[self._login_ui_relation_name]:
+            try:
+                login_ui_endpoints = self.login_ui_endpoints.get_login_ui_endpoints()
+                login_ui_url = login_ui_endpoints[key]
```
Comment (gruyaume) [inline src/charm.py:515]: Place this line outside of the try/except since it's not related to what could raise

Comment (gruyaume) [inline src/charm.py:515]: Here we could simply `return login_ui_endpoints[key]` and not set the `login_ui_url` variable

File: src/charm.py [hunk 6]
```diff
@@ -180,9 +191,9 @@ def _render_conf_file(self) -> str:
 
         rendered = template.render(
             db_info=self._get_database_relation_info(),
-            consent_url=join(self.config.get("login_ui_url", ""), "consent"),
-            error_url=join(self.config.get("login_ui_url", ""), "oidc_error"),
-            login_url=join(self.config.get("login_ui_url", ""), "login"),
+            consent_url=self._get_login_ui_endpoint_info("consent_url"),
```
Comment (gruyaume) [inline src/charm.py:194]: This would be called even if the `ui-endpoint-info` relation doesn't exist it seems, meaning that we'd get `consent_url=None`, `error_url=None`, etc. Is this wanted?

File: src/charm.py [hunk 7]
```diff
@@ -496,6 +507,18 @@ def _update_endpoint_info(self) -> None:
             scope=" ".join(SUPPORTED_SCOPES),
         )
 
+    def _get_login_ui_endpoint_info(self, key: str) -> Optional[str]:
+        login_ui_url = ""
```
[the remark this answers is not in the document]
Comment (bencekov) [reply src/charm.py]: done

**Comment by natalian98** [comment]
Could you add unit tests for your latest changes? The config file seems to be rendered with `None` values (see [logs](https://github.com/canonical/hydra-operator/actions/runs/4611084147/jobs/8150344940?pr=47#step:9:28)) instead of getting the charm in `WaitingStatus`.
Status handling seems to be too complex now, unit tests should help to simplify it.

**Comment by natalian98** [comment]
@bencekov could you solve the conflicts so the CI tests can run?

File: src/charm.py [hunk 8]
```diff
@@ -105,6 +111,10 @@ def __init__(self, *args: Any) -> None:
         )
         self.oauth = OAuthProvider(self)
 
+        self.login_ui_endpoints = LoginUIEndpointsRequirer(
+            self, relation_name=self._login_ui_relation_name
+        )
+
         self.endpoints_provider = HydraEndpointsProvider(self)
 
         self.framework.observe(self.on.hydra_pebble_ready, self._on_hydra_pebble_ready)
```
Comment (natalian98) [inline src/charm.py:120]: Since we removed `config.yaml` completely, we can also stop observing the related event :`self.framework.observe(self.on.config_changed, self._on_config_changed)`

File: src/charm.py [hunk 9]
```diff
@@ -152,6 +162,10 @@ def __init__(self, *args: Any) -> None:
         )
         self.framework.observe(self.on.rotate_key_action, self._on_rotate_key_action)
 
+        self.framework.observe(
+            self.on[self._login_ui_relation_name].relation_changed, self._on_config_changed
```
Comment (natalian98) [inline src/charm.py:166]: Let's not run a method that was used on config changed event, but just `_handle_status_update_config` instead

File: src/charm.py
Comment (natalian98) [inline src/charm.py:1]: Hydra needs some of the login ui endpoints to work properly, e.g. it [can't get the clients](https://github.com/canonical/hydra-operator/actions/runs/4634551274/jobs/8201111771?pr=47#step:8:276) without those urls, which would cause some errors when interacting with hydra CLI. Let's add a default url in case there is no relation or relation data is unavailable, like in kratos.

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -649,6 +697,48 @@ def test_exec_error_on_client_deleted_event_emitted(
...
+                "issuer": "http://127.0.0.1:4444/",
+                "public": "http://127.0.0.1:4444/",
+            },
+        },
+        "webfinger": {
+            "oidc_discovery": {"supported_scope": ["openid", "profile", "email", "phone"]}
+        },
+    }
+
+    assert yaml.safe_load(harness.charm._render_conf_file()) == expected_config
```
Comment (natalian98) [inline tests/unit/test_charm.py:739]: We could compare the expected config with the actual file that was pushed to the container instead of checking `_render_conf_file` method output.
```suggestion
    container_config = container.pull(path="/etc/config/hydra.yaml", encoding="utf-8")
    assert yaml.safe_load(container_config.read()) == expected_config
```

File: lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py
Comment (natalian98) [inline lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py:1]: Please update this lib once the comments in the `identity-platform-login-ui-operator` are addressed and resolved

File: tests/unit/test_charm.py [hunk 11]
```diff
@@ -175,15 +219,15 @@ def test_update_container_config(harness: Harness, mocked_sql_migration: MagicMo
 
 def test_on_config_changed_without_service(harness: Harness) -> None:
     setup_postgres_relation(harness)
-    harness.update_config({"login_ui_url": "http://some-url"})
+    setup_login_ui_relation(harness)
```
Comment (natalian98) [inline tests/unit/test_charm.py:222]: Let's keep the unit tests as minimal as possible. Since the login ui relation is not mandatory for hydra to start (because you set up default urls), I'd simply remove this line
```suggestion

```

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -175,15 +219,15 @@ def test_update_container_config(harness: Harness, mocked_sql_migration: MagicMo
...
 
     assert harness.charm.unit.status == WaitingStatus("Waiting to connect to Hydra container")
 
 
 def test_on_config_changed_without_database(harness: Harness) -> None:
     harness.set_can_connect(CONTAINER_NAME, True)
     harness.charm.on.hydra_pebble_ready.emit(CONTAINER_NAME)
-    harness.update_config({"login_ui_url": "http://some-url"})
+    setup_login_ui_relation(harness)
```
Comment (natalian98) [inline tests/unit/test_charm.py:230]: Similar comment as above, let's keep the test focused on the missing db relation
```suggestion

```

File: tests/unit/test_charm.py [hunk 13]
```diff
@@ -258,6 +314,19 @@ def test_config_updated_on_ingress_relation_joined(harness: Harness) -> None:
...
+        {
+            "default_url": "http://default-url.com",
+            "error_url": "http://default-url.com/error",
+            "login_url": "http://default-url.com/login",
+            "registration_url": "http://default-url.com/registration",
+            "consent_url": "http://default-url.com/consent",
+            "oidc_error_url": "http://default-url.com/oidc_error",
+            "index_url": "http://default-url.com/index",
+        }
+    )
```
Comment (natalian98) [inline tests/unit/test_charm.py:329]: Let's keep this test focused only on the ingress relation if possible

File: templates/hydra.yaml.j2 [hunk 14]
```diff
@@ -18,9 +18,15 @@ serve:
...
 urls:
+  {%- if consent_url %}
   consent: {{ consent_url }}
+  {%- endif %}
+  {%- if error_url %}
   error: {{ error_url }}
+  {%- endif %}
+  {%- if login_url %}
   login: {{ login_url }}
+  {%- endif %}
```
Comment (natalian98) [inline templates/hydra.yaml.j2:29]: Since we have the default urls set up even if the relation doesn't exist and thus those urls will always have some value provided, we don't need the conditions

File: src/charm.py [hunk 15]
```diff
@@ -60,6 +64,13 @@
 HYDRA_PUBLIC_PORT = 4444
 SUPPORTED_SCOPES = ["openid", "profile", "email", "phone"]
 PEER = "hydra"
+LOGIN_UI_DEFAULT_URLS = (
+    {
+        "consent_url": "http://default-url.com/consent",
+        "login_url": "http://default-url.com/login",
+        "oidc_error_url": "http://default-url.com/oidc_error",
+    },
+)
```
Comment (natalian98) [inline src/charm.py:73]: this could also be set in the template like you do in your kratos PR

File: templates/hydra.yaml.j2 (hunk 14 above)
Comment (natalian98) [reply templates/hydra.yaml.j2:29]: dismiss this comment if you decide to go with [this](https://github.com/canonical/kratos-operator/pull/49/files#diff-47cd73aae729ed107a6f43de89a621636ce35f09aa20c357b362a10d8b5f848bR14) approach

File: tests/unit/test_charm.py [hunk 16]
```diff
@@ -649,6 +687,202 @@ def test_exec_error_on_client_deleted_event_emitted(
...
+
+
+def test_config_updated_with_multiple_login_ui_endpoints_interface(
+    harness: Harness, mocked_sql_migration: MagicMock
+) -> None:
+    harness.set_can_connect(CONTAINER_NAME, True)
+    harness.charm.on.hydra_pebble_ready.emit(CONTAINER_NAME)
+    setup_postgres_relation(harness)
+    setup_login_ui_relation(harness)
+    setup_login_ui_relation(harness)
```
Comment (natalian98) [inline tests/unit/test_charm.py:843]: why is it set up twice?

**Review by nsklikas — APPROVED** [review]

**Review by natalian98 — APPROVED** [review]
Please use squash and merge option while merging this PR.

**Review by nsklikas — APPROVED** [review]

**Review by nsklikas — APPROVED** [review]
